prototype/openapiv3/main.go

```diff
@@ -0,0 +1,219 @@
+	if err = json.Unmarshal(data, &openapiv3); err != nil {
+		log.Fatalln(err)
+	}
+
+	/*
+		modify routes
+	*/
+	for name, path := range openapiv3.Paths.Paths {
+		var tags []string
+		switch {
+		case strings.Contains(name, "imagebuilds"):
+			tags = []string{"ImageBuildService"}
+		case strings.Contains(name, "imagecaches"):
+			tags = []string{"ImageCacheService"}
+		}
+
+		if path.Get != nil {
+			mutateOperation(path.Get)
+			path.Get.Tags = tags
+		}
+	}
+
+	for name, path := range openapiv3.Paths.Paths {
+		var tags []string
+		switch {
+		case strings.Contains(name, "imagebuilds"):
+			tags = []string{"ImageBuildService"}
+		case strings.Contains(name, "imagecaches"):
+			tags = []string{"ImageCacheService"}
+		}
+
+		if path.Get != nil {
+			path.Get.Tags = tags
+		}
+		if path.Post != nil {
+			path.Post.Tags = tags
+		}
+		if path.Put != nil {
+			path.Put.Tags = tags
+		}
+		if path.Delete != nil {
+			path.Delete.Tags = tags
+		}
+		if path.Patch != nil {
+			path.Patch.Tags = tags
+		}
+	}
+
+	defs := hephv1.GetOpenAPIDefinitions(func(path string) spec.Ref {
+		switch {
+		case libRE.MatchString(path):
+			path = libRE.ReplaceAllString(path, "")
+		case k8sRE.MatchString(path):
+			path = k8sRE.ReplaceAllString(path, "")
+		}
+		return spec.MustCreateRef("#/components/schemas/" + common.EscapeJsonPointer(path))
+	})
+
+	schemas := map[string]*spec.Schema{}
+	for name, definition := range defs {
+		def := definition
+		schemas[libRE.ReplaceAllString(name, "")] = &def.Schema
+	}
+	openapiv3.Components.Schemas = schemas
+
+	return &openapiv3
+}
+
+func mutateOperation(op *spec3.Operation) {
+	if reqBody := op.RequestBody; reqBody != nil {
+		for _, mediaType := range reqBody.Content {
+			if schema := mediaType.Schema; schema != nil {
+				schema.Ref = spec.MustCreateRef(pathRef(schema.Ref.String()))
+			}
+		}
+	}
+
+	for _, response := range op.Responses.StatusCodeResponses {
+		for _, mediaType := range response.Content {
+			if schema := mediaType.Schema; schema != nil {
+				schema.Ref = spec.MustCreateRef(pathRef(schema.Ref.String()))
+			}
+		}
+	}
+}
+
+func pathRef(path string) string {
+	path = strings.Replace(path, "com.dominodatalab.hephaestus.v1.", "", 1)
+	return path
+
+	// name = strings.ReplaceAll(name, "github.com/dominodatalab/hephaestus/pkg/api/hephaestus/v1", "")
+	// name = strings.ReplaceAll(name, "k8s.io/apimachinery/pkg/apis/meta/", "")
+	//
+	// return name
+}
+
+// func walkOperation(op *spec3.Operation) {
+// 	if reqBody := op.RequestBody; reqBody != nil {
+// 		for _, mediaType := range reqBody.Content {
+// 			if schema := mediaType.Schema; schema != nil {
+// 				ref := schema.Ref
+//
+// 				switch {
+// 				case strings.
+// 				}
+// 			}
+// 		}
+// 	}
+//
+// 	for _, response := range op.Responses.StatusCodeResponses {
+// 		for _, mediaType := range response.Content {
+// 			if schema := mediaType.Schema; schema != nil {
+// 				schema.Ref = refFunc(schema.Ref.String())
+// 			}
+// 		}
+// 	}
+// }
```

Comment:
Is this going to be used for something in a follow-up?

Reply:
Yeah, the `/openapi/v3` endpoint is much more user-friendly than the v2 one because it returns targeted resources (i.e. `/openapi/v3/apis/hephaestus.dominodatalab.com/v1`) instead of EVERYTHING. It's useful but not critical for this work so I'm going to do that work slowly and deliberately as a separate PR when I have time. These are just my notes for that future work.